### PR TITLE
Implement automatic inference of qid shape for quantum states

### PR DESCRIPTION
**Pull request by kevinsung**
Follow up to #3419 . Addresses #2904 .

### CODE REVIEW

File: cirq/qis/states.py [hunk 1]
```diff
@@ -321,6 +313,231 @@ def _infer_qid_shape_from_dimension(dim: int) -> Tuple[int, ...]:
...
+    """Infer the qid shape of a set of states.
+
+    This is a heuristic that is guaranteed to return a qid shape compatible
+    with all of the given states, but may fail (raising an error) even
+    if there is a unique compatible qid shape. If, after applying the heuristic,
+    the qid shape is ambiguous, then an error is raised unless the qid shape is
+    to be inferred from the shape of a state vector or density matrix. In that
+    case, if the dimension of the state space is a power of 2, then the state
+    space is assumed to be composed of qubits; otherwise, it is assumed to be
+    composed of a single qudit.
```
Comment (mpharrigan) [inline cirq/qis/states.py:326]: the syntax of this docstring is hard to parse. It's written like .. backwards .. for computers instead of for humans.

"We use a heuristic to determine a qid shape compatible with all of the given shapes. The heuristic [insert brief description here]. If the dimension of state vector or density matrix (not tensors!) is a power of 2, then we assume qubits, otherwise it is assumed to be composed of a single qudit. If the qid shape is still ambiguous, an error is raised."

File: cirq/qis/states.py [hunk 2]
```diff
@@ -321,6 +313,231 @@ def _infer_qid_shape_from_dimension(dim: int) -> Tuple[int, ...]:
     return (dim,)
 
 
+def infer_qid_shape(*states: 'cirq.QUANTUM_STATE_LIKE') -> Tuple[int, ...]:
+    """Infer the qid shape of a set of states.
+
+    This is a heuristic that is guaranteed to return a qid shape compatible
```
Comment (mpharrigan) [inline cirq/qis/states.py:319]: I think we have different definitions of guarantee. 60% of the time it works every time

File: cirq/qis/states.py [hunk 3]
```diff
@@ -321,6 +313,231 @@ def _infer_qid_shape_from_dimension(dim: int) -> Tuple[int, ...]:
...
+    non_integer_states = []
+    for state in states:
+        if isinstance(state, int):
+            integer_states.append(state)
+        else:
+            non_integer_states.append(state)
+
+    # if all states are specified as integers, the shape is ambiguous
+    if not non_integer_states:
+        raise ValueError('The qid shape of the given states is ambiguous.')
```
Comment (mpharrigan) [inline cirq/qis/states.py:348]: explain why in this error message (not just the comment)

File: cirq/qis/states.py [hunk 4]
```diff
@@ -321,6 +313,231 @@ def _infer_qid_shape_from_dimension(dim: int) -> Tuple[int, ...]:
...
+
+    # attempt to infer qid shape from non-integer states
+    potential_shapes = _potential_qid_shapes(non_integer_states[0])
+    for state in non_integer_states[1:]:
+        potential_shapes = potential_shapes.intersection_subset(_potential_qid_shapes(state))
+    if potential_shapes.is_ambiguous():
+        raise ValueError('The qid shape of the given states is ambiguous.')
```
Comment (mpharrigan) [inline cirq/qis/states.py:355]: change `is_ambiguous` to `assert_ambiguous` and give more helpful errors why it's ambiguous, or change `is_ambiguous` to return a `reason` string, which you can report here

File: cirq/qis/states.py [hunk 5]
```diff
@@ -321,6 +313,231 @@ def _infer_qid_shape_from_dimension(dim: int) -> Tuple[int, ...]:
...
+        return _QidShapeSet(
+            explicit_qid_shapes=explicit_qid_shapes,
+            unfactorized_total_dimension=unfactorized_total_dimension,
+            min_qudit_dimensions=min_qudit_dimensions,
+        )
+
+    def is_ambiguous(self) -> bool:
+        """Whether the qid shape is ambiguous and cannot be inferred."""
+        if self.min_qudit_dimensions is not None:
+            return True
```
Comment (mpharrigan) [inline cirq/qis/states.py:491]: return reasons instead of just true/false so people can figure out what's going on

File: cirq/qis/states.py [hunk 6]
```diff
@@ -321,6 +313,231 @@ def _infer_qid_shape_from_dimension(dim: int) -> Tuple[int, ...]:
...
+
+    # attempt to infer qid shape from non-integer states
+    potential_shapes = _potential_qid_shapes(non_integer_states[0])
+    for state in non_integer_states[1:]:
+        potential_shapes = potential_shapes.intersection_subset(_potential_qid_shapes(state))
+    if potential_shapes.is_ambiguous():
+        raise ValueError('The qid shape of the given states is ambiguous.')
+    qid_shape = potential_shapes.infer_qid_shape()
+    if qid_shape is None:
+        raise ValueError('Failed to infer the qid shape of the given states.')
```
Comment (mpharrigan) [inline cirq/qis/states.py:358]: more descriptive error telling you why. there's also a redundant check, since `infer_qid_shape` calls `is_ambiguous`

File: cirq/qis/states.py [hunk 7]
```diff
@@ -321,6 +313,231 @@ def _infer_qid_shape_from_dimension(dim: int) -> Tuple[int, ...]:
...
+
+    # check if the shape is compatible with the states specified as integers
+    if integer_states and np.prod(qid_shape, dtype=int) <= max(integer_states):
+        raise ValueError('Failed to infer the qid shape of the given states.')
```
Comment (mpharrigan) [inline cirq/qis/states.py:362]: describe the failure!

File: cirq/qis/states.py [hunk 8]
```diff
@@ -321,6 +313,231 @@ def _infer_qid_shape_from_dimension(dim: int) -> Tuple[int, ...]:
...
+
+    Returns:
+        The inferred qid shape.
+
+    Raises:
+        ValueError: The qid shape of the given states is ambiguous.
+        ValueError: Failed to infer the qid shape of the given states.
+    """
+    integer_states = []
+    non_integer_states = []
```
Comment (mpharrigan) [inline cirq/qis/states.py:339]: type annotations could be helpful here

File: cirq/qis/states.py [hunk 9]
```diff
@@ -321,6 +313,231 @@ def _infer_qid_shape_from_dimension(dim: int) -> Tuple[int, ...]:
...
+    np.ndarray,
+    Sequence[Union[int, float, complex]],
+    # Product state object
+    'cirq.ProductState',
+    # Quantum state object
+    'cirq.QuantumState',
+]
+
+
+def _potential_qid_shapes(state: _NON_INT_STATE_LIKE) -> '_QidShapeSet':
```
Comment (mpharrigan) [inline cirq/qis/states.py:380]: short docstring would be helpful here

File: cirq/qis/states.py [hunk 10]
```diff
@@ -321,6 +313,231 @@ def _infer_qid_shape_from_dimension(dim: int) -> Tuple[int, ...]:
...
+
+    if isinstance(state, Sequence):
+        state = np.array(state)
+    if state.ndim == 1:
+        (dim,) = state.shape
+        if state.dtype.kind == 'c':
+            # definitely state vector amplitudes
+            return _QidShapeSet(unfactorized_total_dimension=dim)
+        # could also be per-qudit computational basis values
+        min_qudit_dimensions = tuple(state.astype(int, copy=False) + 1)
```
Comment (mpharrigan) [inline cirq/qis/states.py:394]: what if `state.dtype.kind` is float (i.e. it just happens to not have complex elements)

File: cirq/qis/states.py [hunk 11]
```diff
@@ -321,6 +313,231 @@ def _infer_qid_shape_from_dimension(dim: int) -> Tuple[int, ...]:
...
+            explicit_qid_shapes |= _intersection_explicit_with_unfactorized_qid_shapes(
+                other.explicit_qid_shapes, self.unfactorized_total_dimension
+            )
+
+        if self.unfactorized_total_dimension == other.unfactorized_total_dimension:
+            unfactorized_total_dimension = self.unfactorized_total_dimension
+
+        # don't attempt to intersect unfactorized total dimension with
+        # min qudit dimensions
+
```
Comment (mpharrigan) [inline cirq/qis/states.py:468]: extra newline?

File: cirq/qis/states.py [hunk 12]
```diff
@@ -321,6 +313,231 @@ def _infer_qid_shape_from_dimension(dim: int) -> Tuple[int, ...]:
...
+        if self.min_qudit_dimensions is not None and other.explicit_qid_shapes:
+            explicit_qid_shapes |= _intersection_explicit_with_min_qudit_dims_qid_shapes(
+                other.explicit_qid_shapes, self.min_qudit_dimensions
+            )
+
+        # don't attempt to intersect min qudit dimensions with
+        # unfactorized total dimension
+
```
Comment (mpharrigan) [inline cirq/qis/states.py:476]: extra newline?

**Review by mpharrigan — COMMENTED** [review]
cool

File: cirq/qis/states.py (hunk 1 above)
Comment (kevinsung) [reply cirq/qis/states.py]: I've rewritten it following your suggestion.

File: cirq/qis/states.py (hunk 3 above)
Comment (kevinsung) [reply cirq/qis/states.py]: Done.

File: cirq/qis/states.py (hunk 4 above)
Comment (kevinsung) [reply cirq/qis/states.py]: Done.

File: cirq/qis/states.py (hunk 5 above)
Comment (kevinsung) [reply cirq/qis/states.py]: Done.

File: cirq/qis/states.py (hunk 6 above)
Comment (kevinsung) [reply cirq/qis/states.py]: Done.

File: cirq/qis/states.py (hunk 7 above)
Comment (kevinsung) [reply cirq/qis/states.py]: Done.

File: cirq/qis/states.py (hunk 8 above)
Comment (kevinsung) [reply cirq/qis/states.py]: Done.

File: cirq/qis/states.py (hunk 9 above)
Comment (kevinsung) [reply cirq/qis/states.py:401]: Done.

File: cirq/qis/states.py (hunk 10 above)
Comment (kevinsung) [reply cirq/qis/states.py]: I've changed the code to check the type before converting to int.

File: cirq/qis/states.py (hunk 11 above)
Comment (kevinsung) [reply cirq/qis/states.py]: Those were deliberate, but I've changed these comments to be more clear.

File: cirq/qis/states.py (hunk 12 above)
Comment (kevinsung) [reply cirq/qis/states.py]: Those were deliberate, but I've changed these comments to be more clear.

**Comment by kevinsung** [comment]
@mpharrigan I've addressed your comments, please have another look.

File: cirq/qis/measures.py [hunk 13]
```diff
@@ -132,10 +133,7 @@ def fidelity(
 
     # Use QuantumState machinery for the general case
     if qid_shape is None:
-        # coverage: ignore
-        raise NotImplementedError(
-            'Qid shape inference not yet implemented. Please specify the qid shape explicitly.'
-        )
+        qid_shape = infer_qid_shape(state1, state2)
```
Comment (balopat) [inline cirq/qis/measures.py:136]: I think a message to specify the qid shape explicitly is still useful here to help the user. Otherwise only the inference errors will be shown which are a bit more obscure.

File: cirq/__init__.py [hunk 14]
```diff
@@ -331,6 +331,7 @@
     dirac_notation,
     eye_tensor,
     fidelity,
+    infer_qid_shape,
```
Comment (balopat) [inline cirq/__init__.py:334]: I'm not convinced yet that this needs to be exposed to the cirq namespace. 
It feels like this would be useful mostly for other functions in the qis package.

**Comment by balopat** [comment]
A couple of comments: 
- There is also `cirq.to_valid_state_vector` that should make use of this functionality - `states._state_like_to_state_tensor` seems to do a bunch of qid shape inference itself...

After playing around with the code, it felt a bit fragile and hard to understand the rules just by discovery...it might be that I'm not smart enough for it or I'm misusing the tool :) 

- I wonder if we want to adopt a "use minimal dimensions where possible" - it would allow for `infer_qid_shape(1)` to be `(2,1)`
- I wonder if we want to adopt automated casting - e.g. `infer_qid_shape([1,0])` fails now but  `infer_qid_shape([1,1j])` doesn't...what do we think about automated casting to complex?

**Comment by mpharrigan** [comment]
> I wonder if we want to adopt automated casting - e.g. infer_qid_shape([1,0]) fails now but infer_qid_shape([1,1j]) doesn't...what do we think about automated casting to complex?

I think this is by design to allow integer specification of basis states vs complex amplitudes

File: cirq/qis/states.py [hunk 15]
```diff
@@ -343,39 +368,36 @@ def infer_qid_shape(*states: 'cirq.QUANTUM_STATE_LIKE') -> Tuple[int, ...]:
...
-    if potential_shapes.is_ambiguous():
-        raise ValueError('The qid shape of the given states is ambiguous.')
     qid_shape = potential_shapes.infer_qid_shape()
     if qid_shape is None:
-        raise ValueError('Failed to infer the qid shape of the given states.')
+        raise ValueError(
+            'Failed to infer the qid shape of the given states. '
+            'Either there is no compatible qid shape for the given states, or there is one '
+            'but the heuristic failed to find it.'
+        )
```
Comment (mpharrigan) [inline cirq/qis/states.py:386]: The rest of the new error messages look great. This one reads really funny, however. "Either there is none or there is one" is a bad excuse for a program to make :) :) If this is truly an edge case, I'd recommend admitting as much

> Failed to infer the qid shape of the given states. If you think there is an unambiguous result for this case, please raise a GitHub issue.

Otherwise, parcel out the conditionals that can lead here: e.g. 

> Could not determine the total (unfactorized) Hilbert space dimension and could not find any explicit Qid shapes.

File: cirq/__init__.py (hunk 14 above)
Comment (mpharrigan) [reply cirq/__init__.py:334]: We've never really held back from exposing every public function to the global namespace before. We've never had a good packaging strategy before either. But in any event, we should make the criteria clear for global namespace instead of picking and choosing per PR

**Review by mpharrigan — APPROVED** [review]
My comments have been addressed. Please look at @balopat 's tho

**Comment by balopat** [comment]
> > I wonder if we want to adopt automated casting - e.g. infer_qid_shape([1,0]) fails now but infer_qid_shape([1,1j]) doesn't...what do we think about automated casting to complex?
> 
> I think this is by design to allow integer specification of basis states vs complex amplitudes

I'm confused - list of ints wouldn't be integer specification of basis states as far as I understand, `infer_qid_shape([1,0])` is the `|0>` state, it's just not complex...

**Comment by balopat** [comment]
> > > I wonder if we want to adopt automated casting - e.g. infer_qid_shape([1,0]) fails now but infer_qid_shape([1,1j]) doesn't...what do we think about automated casting to complex?
> > 
> > 
> > I think this is by design to allow integer specification of basis states vs complex amplitudes
> 
> I'm confused - list of ints wouldn't be integer specification of basis states as far as I understand, `infer_qid_shape([1,0])` is the `|0>` state, it's just not complex...

Oh, I got it - it could be looked at as a multiqubit integer state `|10>`...ahhh...

File: cirq/__init__.py (hunk 14 above)
Comment (mpharrigan) [reply cirq/__init__.py:334]: I'm always in favor of making things package-private and expanding as necessary

File: cirq/qis/measures.py (hunk 13 above)
Comment (kevinsung) [reply cirq/qis/measures.py]: Done.

File: cirq/__init__.py (hunk 14 above)
Comment (kevinsung) [reply cirq/__init__.py]: I've removed the top-level exposure.

File: cirq/qis/states.py [hunk 16]
```diff
@@ -311,17 +323,30 @@ def _infer_qid_shape_from_dimension(dim: int) -> Tuple[int, ...]:
...
-    space is assumed to be composed of qubits; otherwise, it is assumed to be
-    composed of a single qudit.
+    This is a heuristic to determine a qid shape compatible with all of the
+    given states. It works by attempting to find the intersection of the sets
+    of potential qid shapes for each given state. It may fail (raising an
+    error) even if there is a unique compatible qid shape. If the dimension of
+    a state vector or density matrix (but not state tensor) is a power of 2,
+    then the state space is assumed to be composed of qubits; otherwise, it is
+    assumed to be composed of a single qudit. If the qid shape is ambiguous,
+    an error is raised.
```
[the remark this answers is not in the document]
Comment (kevinsung) [reply cirq/qis/states.py:349]: I agree, thanks for the suggestion!

File: cirq/qis/states.py (hunk 15 above)
Comment (kevinsung) [reply cirq/qis/states.py]: I took your first suggestion; I do believe that this case would be reached only in unusual circumstances.

**Comment by kevinsung** [comment]
> A couple of comments:
> 
>     * There is also `cirq.to_valid_state_vector` that should make use of this functionality - `states._state_like_to_state_tensor` seems to do a bunch of qid shape inference itself...
> 
> 
> After playing around with the code, it felt a bit fragile and hard to understand the rules just by discovery...it might be that I'm not smart enough for it or I'm misusing the tool :)
> 
>     * I wonder if we want to adopt a "use minimal dimensions where possible" - it would allow for `infer_qid_shape(1)` to be `(2,1)` --- python world precedence: `bin(3) -> 0b11'
> 
>     * I wonder if we want to adopt automated casting - e.g. `infer_qid_shape([1,0])` fails now but  `infer_qid_shape([1,1j])` doesn't...what do we think about automated casting to complex?

I'll edit `states._state_like_to_state_tensor` to use this, and I'll also implement your suggestion to use minimum dimensions for integer states.

**Comment by kevinsung** [comment]
@balopat I've addressed your comments, please have another look.

**Comment by mpharrigan** [comment]
@balopat did Kevin address your comments?

**Review by balopat — APPROVED** [review]
LGTM, thank you!

**Comment by mpharrigan** [comment]
@kevinsung can you sign the CLA?

**Comment by kevinsung** [comment]
@mpharrigan I signed one years ago; do you have any idea why the check would stop going through all of a sudden?

**Comment by balopat** [comment]
All is good on the CLA front, I think either Cirqbot might have a bug here or there was a CLA bot flake? I suspect the former.
